Add new record types in batch change

Aravindh-Raju opened the pull request:
Fixes #1248 and #1255.

Changes in this pull request:
- Add NAPTR, SRV and NS record types in batch change.
- Add support for the above record types through csv.

nspadaccino commented on modules/api/src/main/scala/vinyldns/api/route/BatchChangeJsonProtocol.scala:
```diff
@@ -249,8 +249,23 @@ trait BatchChangeJsonProtocol extends JsonValidation {
+          "Missing BatchChangeInput.changes.record.preference and " +
+          "Missing BatchChangeInput.changes.record.flags and " +
+          "Missing BatchChangeInput.changes.record.service and " +
+          "Missing BatchChangeInput.changes.record.regexp and " +
```
regexp is optional right? Or is it required via the API?

Aravindh-Raju replied:
Regexp was optional, but the field was made required while making NAPTR changes via Zones view as seen in `DNSJsonProtocol.scala` - https://github.com/vinyldns/vinyldns/blob/master/modules/api/src/main/scala/vinyldns/api/route/DnsJsonProtocol.scala#L524-L528

So, I made it required in `BatchChangeJsonProtocol.scala` too. The `regexp` field is required but the data is optional and can be left empty.

nspadaccino commented on modules/docs/src/main/resources/microsite/static/dns-changes-csv-sample.csv:
```diff
@@ -6,6 +6,11 @@ Add,AAAA,test3.example.com.,,fd69:27cc:fe91::60
 Add,TXT,test5.example.com,7200,example text
+Add,MX,test16.ok.,7200,1 text.com
+Add,NS,test17.ok.,7200,172.17.42.1
+Add,NAPTR,test18.ok.,7200,10 20 s SIPS+D2T _si._tc.ac-ot1.wdv.test.net.
```
So in the case where the record data has multiple fields like for NAPTR and SRV, the values should be space separated?

Aravindh-Raju replied:
Yes, in the case where the record data has multiple fields, the values are space separated. But if we want to have some other delimiter instead of space, I can do that.

nspadaccino reviewed (changes requested):
Tested out locally and the functionality looks good. Just one thing: for NAPTR records, the only validation we have on the Flags input is that it should be less than two characters. I am able to submit flags other than U, S, A, or P (such as "F" or "1"), but when I do, the record change fails with no message. I think if those flags are the only valid ones, we should validate the input to make sure the flag is one of those

Aravindh-Raju commented:
> Tested out locally and the functionality looks good. Just one thing: for NAPTR records, the only validation we have on the Flags input is that it should be less than two characters. I am able to submit flags other than U, S, A, or P (such as "F" or "1"), but when I do, the record change fails with no message. I think if those flags are the only valid ones, we should validate the input to make sure the flag is one of those

Added validation for NAPTR flags in api and converted the text box to dropdown in portal with U, S, A, and P values.

nspadaccino commented on modules/docs/src/main/resources/microsite/static/dns-changes-csv-sample.csv:
```diff
@@ -6,6 +6,11 @@ Add,AAAA,test3.example.com.,,fd69:27cc:fe91::60
+Add,MX,test16.ok.,7200,1 text.com
+Add,NS,test17.ok.,7200,172.17.42.1
+Add,NAPTR,test18.ok.,7200,10 20 s SIPS+D2T _si._tc.ac-ot1.wdv.test.net.
+Add,NAPTR,test20.ok.,7200,10 20 s SIPS+D2T !^.*$!sip:jd@corpxyz.com! _si._tc.ac-ot1.wdv.test.net.
```
With the new validation for the Flags field, the flags have to be uppercase if you are importing from csv, so we should change that here

Aravindh-Raju replied:
Made the changes

nspadaccino commented:
Another thing I noticed is that if you enter a string in the regex field and the string does not begin and end with an "!", the record change passes validation but fails, so I think we should add some input validation for the regex field

Aravindh-Raju commented:
> Another thing I noticed is that if you enter a string in the regex field and the string does not begin and end with an "!", the record change passes validation but fails, so I think we should add some input validation for the regex field

Looks like regexp should begin and end with '!' or can be empty. Added the validation now.

nspadaccino commented:
Could we also enforce the input validation when creating NAPTR records from the zone view? Unless you think that would be better in another PR since this PR is about batch changes

Aravindh-Raju commented:
> Could we also enforce the input validation when creating NAPTR records from the zone view? Unless you think that would be better in another PR since this PR is about batch changes

I think it would be better to open a separate PR, incase we need to refer that individually in the future. I'll open one soon with the changes.

nspadaccino reviewed (approved):
Looks good to me
